File: algos/range_queries/PrefixSum/soln/Aditya Pratap Singh/Solution2.cpp
/*
------------------------------------------------------------
Problem Statement: Running Sum of 1D Array (LeetCode 1480)

Short Problem Statement:
Given an array nums, return the running sum of the array,
where runningSum[i] is the sum of all elements from index
0 to i.

------------------------------------------------------------
Approach (Using Prefix Sum):

We construct a prefix sum array where:
prefixSum[i] = prefixSum[i - 1] + nums[i]

The first element remains the same since it has no
previous elements.

------------------------------------------------------------
Time Complexity:
- O(n), where n is the size of the array

Space Complexity:
- O(n) for the prefix sum array

------------------------------------------------------------
Example:
Input:
nums = [1, 2, 3, 4]

Output:
[1, 3, 6, 10]

------------------------------------------------------------
*/

#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    vector<int> runningSum(vector<int> &nums)
    {
        int n = nums.size();
        vector<int> prefixSumArray(n);

        prefixSumArray[0] = nums[0];
        for (int i = 1; i < n; i++)
        {
            prefixSumArray[i] = prefixSumArray[i - 1] + nums[i];
        }

        return prefixSumArray;
    }
};

File: algos/range_queries/PrefixSum/soln/verifiedHuman18/Solution2.java
/*
PROBLEM STATEMENT:
Given an array nums. We define a running sum of an array as runningSum[i] = sum(nums[0]…nums[i]).
Return the running sum of nums.
*/

/*
Explanation:
For a given array of integers, we create a prefix sum array that keeps store of cumulative sum of all elements upto the specific index.
So for an array [a, b, c], the prefix becomes
[a, a + b, a + b + c]
*/

class Solution2 {
    public static int[] runningSum(int[] nums) {
        int[] prefix = new int[nums.length];
        prefix[0] = nums[0];
        for (int i = 1; i < nums.length; i++) {
            prefix[i] = prefix[i - 1] + nums[i];
        }  
        return prefix;
    }

    public static void main(String[] args) {
        int[] arr = {-2, 0, 3, -5, 2, -1};
        int[] sumArr = runningSum(arr);
        for (int i = 0; i < sumArr.length; i++) {
            System.out.print(sumArr[i] + " ");
        }
        
    }
}

// Time Complexity = O(n)
// Space Complexity = O(n)